backward.hpp

```diff
@@ -515,12 +587,19 @@ class StackTraceLinuxImplBase {
 
 protected:
 	void load_thread_info() {
+#ifdef BACKWARD_SYSTEM_LINUX
 		_thread_id = (size_t)syscall(SYS_gettid);
 		if (_thread_id == (size_t) getpid()) {
 			// If the thread is the main one, let's hide that.
 			// I like to keep little secret sometimes.
 			_thread_id = 0;
 		}
+#elif defined(BACKWARD_SYSTEM_DARWIN)
+		uint64_t tid;
+		pthread_threadid_np(NULL, &tid);
+
+		_thread_id = static_cast<unsigned>(tid);
```

Comment:
you are casting 64bits to 32bits. looks like a bad idea. pthread_threadid_np returns a system wide unique thread id. Which is macos specific and definitively not needed here.

The main reason I used gettid in the first place, was to conveniently identify the main thread (useful info when printing the trace).

With pthread, the unique to the current process id is the pthread_t type itself. Its an opaque type that can be obtained by calling pthread_self.

Technically on linux, pthread is not the only possible thread implementation. Which means gettid has the advantage of being independent of pthread. Calling pthread functions means we force the user of backward to link with the pthread library.

Macos/bsd pthread have pthread_main_np which tells you if you are in the main thread.

So we have a way to identify the main thread on linux and mac. And we canse use pthread_t for the thread identifier. But again, this makes backward dependent on pthread.

I will continue thinking of the problem and report my finding.

Reply:
This could be rewritten in terms of some new pre-preprocessor definitions, say `BACKWARD_HAS_GETTID` and `BACKWARD_HAS_PTHREADS`. Linux could default to the former, and MacOS to the latter. Linux consumers could optionally select the pthread implementation if they desired it.

Reply:
Updated the implementation for MacOS to use pthread_self and pthread_main_np. Since pthread seems to be always available on MacOS, I'm not sure there's any reason to avoid using it. I misunderstood your comment and didn't immediately realize that pthread_main_np isn't available on Linux. In that case, I don't think it's worth providing a Linux pthread implementation since gettid/getpid works.